### ISSUE TITLE
Strange memory corruption when doing numerics.

### ISSUE THREAD

**Issue by gambiteer** [gambit/gambit#737]
I configured and built Gambit with
```
heine:~/programs/gambit/gambit> gsi/gsi -v
v4.9.4-16-g05b050f8 20220210003317 x86_64-pc-linux-gnu "./configure '--enable-debug'"
```
I started with this (old, incorrect) definition of `partial-factorial`:
```
(define (partial-factorial m n)
  ;; computes the product (m+1) * ... * (n-1) * n
  (if (< (- n m) 10)
      (apply * (iota (+ m 1) (+ n 1)))
      (* (partial-factorial m (quotient (+ m n) 2))
         (partial-factorial (quotient (+ m n) 2) n))))q
```
And, I'm getting very strange errors that are partially reproducible.

A "typical" session is attached. (Sorry, there's a large number printed in the REPL, but I wanted to exactly show the session where things went wrong.)  The problem may be in the numerics (of course), but somehow it seems more general than that.

I'll continue to look at it, but any pointers would be helpful.

Brad
[error-output.txt](https://github.com/gambit/gambit/files/8120199/error-output.txt)

**Comment by feeley** [comment]
The call to `apply` is probably the cause of the issue.  The numerically correct call is `(apply * (iota (- n m) (+ m 1)))` which you probably know already.  The (suspected) cause of the issue is that the call `(iota (+ m 1) (+ n 1))` is creating up to a 10,000 element list which is beyond the implementation limit of 8192 arguments.  I'll have to investigate why in this particular code the limit is not being detected properly.  You could verify this by increasing the limit in `mem.h` and trying to run the same program again:

```
#define ___MAX_NB_ARGS          20000 /* was 8192 */
```

**Comment by gambiteer** [comment]
No, this hasn’t hit the 8192 argument limit yet, in another test when it did, the limit was flagged correctly.

**Comment by feeley** [comment]
It may be below the limit but not far away, so my guess is that things go wrong when the number of arguments passed to `apply` is close to the limit.  By increasing the limit the runtime system will have reserved more space for the arguments, so this may be enough to avoid the issue.

**Comment by feeley** [comment]
Another idea is to avoid `apply` by replacing `(apply * (iota (- n m) (+ m 1)))` by `(fold * 1 (iota (- n m) (+ m 1)))` to see if the problem originates from `apply`.

**Comment by gambiteer** [comment]
Good idea, this didn't error:
```
(define (partial-factorial2 m n)
  ;; computes the product (m+1) * ... * (n-1) * n
  (if (< (- n m) 10)
      (fold * 1 (iota (+ m 1) (+ n 1)))
      (* (partial-factorial2 m (quotient (+ m n) 2))
         (partial-factorial2 (quotient (+ m n) 2) n))))
```

**Comment by gambiteer** [comment]
Re: "my guess is that things go wrong when the number of arguments passed to apply is close to the limit."
Maybe.  It failed with about 6500 arguments.
